applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp

```diff
@@ -0,0 +1,76 @@
+// KRATOS___
+//     //   ) )
+//    //         ___      ___
+//   //  ____  //___) ) //   ) )
+//  //    / / //       //   / /
+// ((____/ / ((____   ((___/ /  MECHANICS
+//
+//  License:         geo_mechanics_application/license.txt
+//
+//  Main authors:    Mohamed Nabi
+//                   John van Esch
+//                   Gennady Markelov
+//
+
+#include "custom_constitutive/thermal_dispersion_law.h"
+#include "custom_retention/retention_law_factory.h"
+#include "geo_mechanics_application_variables.h"
+
+namespace Kratos {
+
+GeoThermalDispersionLaw::GeoThermalDispersionLaw(std::size_t NumberOfDimensions)
+    : mNumberOfDimensions{NumberOfDimensions}
+{
+    KRATOS_ERROR_IF(mNumberOfDimensions < 2 || mNumberOfDimensions > 3)
+        << "Got invalid number of dimensions: " << mNumberOfDimensions << std::endl;
+}
+
+ConstitutiveLaw::Pointer GeoThermalDispersionLaw::Clone() const
+{
+    return Kratos::make_shared<GeoThermalDispersionLaw>(*this);
+}
+
+SizeType GeoThermalDispersionLaw::WorkingSpaceDimension()
+{
+    return mNumberOfDimensions;
+}
+
+Matrix GeoThermalDispersionLaw::CalculateThermalDispersionMatrix(
+    const Properties& rProp, const ProcessInfo& rProcessInfo, const Geometry<Node>& rElementGeometry) const
+{
+    KRATOS_TRY
+
+    Matrix result = ZeroMatrix(mNumberOfDimensions, mNumberOfDimensions);
+
+    RetentionLaw::Parameters parameters(rElementGeometry, rProp, rProcessInfo);
+    auto retention_law = RetentionLawFactory::Clone(rProp);
+    const double saturation = retention_law->CalculateSaturation(parameters);
+    const double water_fraction = rProp[POROSITY] * saturation;
+    const double solid_fraction = 1.0 - rProp[POROSITY];
+
+    const int x = static_cast<int>(indexThermalFlux::X);
+    const int y = static_cast<int>(indexThermalFlux::Y);
```

Comment:
I would replace the explicit type (`int`) by `auto`, since it is repeating the type expressed by the `static_cast<int>`. SonarQube will report code smells for these as well.
```suggestion
    const auto x = static_cast<int>(indexThermalFlux::X);
    const auto y = static_cast<int>(indexThermalFlux::Y);
```

Reply:
changed.